wikipedia_cleanup/data_filter.py

```diff
@@ -257,6 +263,86 @@ def change_pair_needs_to_be_filtered(
         )
 
 
+class FeatureAdderFilter(AbstractDataFilter):
+    def _filter_for_property(self, changes: List[InfoboxChange]) -> List[InfoboxChange]:
+        change_timestamps = pd.Series((change.value_valid_from for change in changes))
+        change_timestamps = (
+            change_timestamps.dt.normalize().dt.normalize().dt.tz_localize(None)
+        )
+        day_of_year = change_timestamps.dt.dayofyear
+        day_of_month = change_timestamps.dt.day
+        day_of_week = change_timestamps.dt.dayofweek
+        month_of_year = change_timestamps.dt.month
+        quarter_of_year = change_timestamps.dt.quarter
+        is_month_start = change_timestamps.dt.is_month_start
+        is_month_end = change_timestamps.dt.is_month_end
+        is_quarter_start = change_timestamps.dt.is_quarter_start
+        is_quarter_end = change_timestamps.dt.is_quarter_end
+
+        days_since_last_change = (
+            change_timestamps
+            - change_timestamps.shift(+1).fillna(pd.Timestamp("20990101"))
+        ).dt.days
+        days_since_last_change = days_since_last_change.copy()
+        days_since_last_change.loc[days_since_last_change < 0] = 0
+
+        days_since_last_2_changes = (
+            change_timestamps
+            - change_timestamps.shift(+2).fillna(pd.Timestamp("20990101"))
+        ).dt.days
+        days_since_last_2_changes = days_since_last_2_changes.copy()
+        days_since_last_2_changes.loc[(days_since_last_2_changes < 0)] = 0
+
+        days_since_last_3_changes = (
+            change_timestamps
+            - change_timestamps.shift(+3).fillna(pd.Timestamp("20990101"))
+        ).dt.days
+        days_since_last_3_changes = days_since_last_3_changes.copy()
+        days_since_last_3_changes.loc[days_since_last_3_changes < 0] = 0
+
+        days_until_next_change = days_since_last_change.shift(-1)
+        days_until_next_change = pd.to_numeric(
+            days_until_next_change.fillna(0), downcast="integer"
+        )
+
+        days_between_last_and_2nd_to_last_change = days_since_last_change.shift(+1)
+        days_between_last_and_2nd_to_last_change = pd.to_numeric(
+            days_between_last_and_2nd_to_last_change.fillna(0), downcast="integer"
+        )
+
+        mean_change_frequency_all_previous = days_since_last_change.mean()
+
+        mean_change_frequency_last_3 = (
+            days_since_last_change.rolling(3).mean().fillna(0)
+        )
```

Comment:
It might be a lot more efficient to port this to numpy. Im also not quite sure if all the copies are really needed, but I'd have to look into it.

Reply:
Yes, some stuff could be converted to numpy. I am just not sure if it is worth the effort. We will probably create a dataset with all features, so we do not to generate them each time again.